### PR TITLE
Null check and output compression

### PR DESCRIPTION
**Pull request by Rolstenhouse**

### CODE REVIEW

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 1]
```diff
@@ -1,4 +1,5 @@
 #Reads in SARIF files and rewrite absolute URIs to reference the source control version
+$ErrorActionPreference = "Stop"
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Yes, that's the incantation I couldn't remember.

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 2]
```diff
@@ -17,16 +18,18 @@ else{
...
-      $caseSensitiveUri = Get-CaseSensitivePath($originalURI.SubString("file:///".Length))
-      $replaceLength = $caseSensitiveUri.IndexOf($projectSlug)
-      if($replaceLength -ne -1){
-          $builder.Path = ($repoName, $repoCommit, $caseSensitiveUri.SubString($replaceLength+$projectSlug.Length+1) -join '/')
-          return $builder.ToString()
-      }
+  if($originalURI){
+    $replace = $originalURI.IndexOf("file:///")
+    if($replace -ne  -1){
+        $caseSensitiveUri = Get-CaseSensitivePath($originalURI.SubString("file:///".Length))
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Remember, the arguments to PowerShell functions are not enclosed in parentheses (for a subtle reason that you can look up if you like). This should be:

```
$caseSensitiveUri = Get-CaseSensitivePath $originalURI.SubString("file:///".Length)
```

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 3]
```diff
@@ -17,16 +18,18 @@ else{
...
-  $replace = $originalURI.IndexOf("file:///")
-  if($replace -ne  -1){
-      $caseSensitiveUri = Get-CaseSensitivePath($originalURI.SubString("file:///".Length))
-      $replaceLength = $caseSensitiveUri.IndexOf($projectSlug)
-      if($replaceLength -ne -1){
-          $builder.Path = ($repoName, $repoCommit, $caseSensitiveUri.SubString($replaceLength+$projectSlug.Length+1) -join '/')
-          return $builder.ToString()
-      }
+  if($originalURI){
+    $replace = $originalURI.IndexOf("file:///")
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Since this literal string occurs twice, let's DRY it out with a variable:

```
$fileUriPrefix = "file:///"
...

... and use the variable both places you use the string.

Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: But also, you only want to do this replacement if that string occurs at the _start_ of the URI. As written, the condition `if ($replace -ne -1)` will be true if that string occurs _anywhere_ in the URI. So change this to:

```
if ($originalUri.StartsWith($fileUriPrefix) {
```

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 4]
```diff
@@ -17,16 +18,18 @@ else{
 }
 
 function Rebase-Uri($originalURI){
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Naming: For consistency with other variables in  the code, and with .NET coding conventions (where acronyms start with a single capital letter), rename this to `$originalUri`.

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 5]
```diff
@@ -17,16 +18,18 @@ else{
...
-      $replaceLength = $caseSensitiveUri.IndexOf($projectSlug)
-      if($replaceLength -ne -1){
-          $builder.Path = ($repoName, $repoCommit, $caseSensitiveUri.SubString($replaceLength+$projectSlug.Length+1) -join '/')
-          return $builder.ToString()
-      }
+  if($originalURI){
+    $replace = $originalURI.IndexOf("file:///")
+    if($replace -ne  -1){
+        $caseSensitiveUri = Get-CaseSensitivePath($originalURI.SubString("file:///".Length))
+        $replaceLength = $caseSensitiveUri.IndexOf($projectSlug)
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: `IndexOf` doesn't return a length, it returns a 0-based index. A nice descriptive name for this variable would be `$projectSlugIndex`.

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 6]
```diff
@@ -50,20 +53,24 @@ function Get-CaseSensitivePath($pathName){
...
-    if($content.uri){
-        $content.uri = Rebase-Uri $content.uri
-    }
-
-    if($content.parentKey){
-      $content.parentKey = Rebase-Uri $content.parentKey
+  if($sarifLog.runs[$i].files){
+    $sarifLog.runs[$i].files.PSObject.Properties | foreach-object{
+      $key = $_.Name
+      $content = $_.Value
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: It's conventional to call these `$key` and `$value`.

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 7]
```diff
@@ -50,20 +53,24 @@ function Get-CaseSensitivePath($pathName){
...
+        if($content.uri){
+            $content.uri = Rebase-Uri $content.uri
+        }
+
+        if($content.parentKey){
+          $content.parentKey = Rebase-Uri $content.parentKey
+        }
+
+        $sarifLog.runs.files.PSObject.Properties.Remove($_.Name)
+        $rewrite = Rebase-Uri $key
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: 1. Variable names are generally nouns; "rewrite" is a verb.
2. You chose the name `Rebase-Uri` for your function (which is a good name), so in any case the verb should be "rebase" rather than "rewrite".

So... how about `$rebasedUri`?

File: appveyor/Convert-FileUrisToGitHubUris.ps1 (hunk 6 above)
Comment (Rolstenhouse) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Fixed

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 8]
```diff
@@ -81,32 +88,50 @@ for ($i = 0; $i -lt $sarifLog.runs.Count; $i++){
...
+      if($uri){
+        $sarifLog.runs[$i].results[$j].stacks[$k].frames.uri = Rebase-Uri $uri
+      }
     }
 
     for($k = 0; $k -lt $sarifLog.runs[$i].results[$j].fixes.Count; $k++){
-      $sarifLog.runs[$i].results[$j].fixes[$k].fileChanges.uri = Rebase-Uri $sarifLog.runs[$i].results[$j].fixes[$k].fileChanges.uri
+      $uri = $sarifLog.runs[$i].results[$j].fixes[$k].fileChanges.uri
+      if($uri){
+      $sarifLog.runs[$i].results[$j].fixes[$k].fileChanges.uri = Rebase-Uri $uri
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Indentation doesn't match (maybe there's a tab in here?).

File: appveyor/Convert-FileUrisToGitHubUris.ps1 (hunk 4 above)
Comment (Rolstenhouse) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Fixed

File: appveyor/Convert-FileUrisToGitHubUris.ps1 (hunk 5 above)
Comment (Rolstenhouse) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Fixed

File: appveyor/Convert-FileUrisToGitHubUris.ps1 (hunk 3 above)
Comment (Rolstenhouse) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Fixed. This was the functionality that was intended

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 9]
```diff
@@ -81,32 +88,50 @@ for ($i = 0; $i -lt $sarifLog.runs.Count; $i++){
...
+    if($uri){
+      $sarifLog.runs[$i].toolNotifications[$j].physicalLocation.uri = Rebase-Uri $uri
+    }
   }
 }
 
 $newExtension = ".github.sarif"
 $writeLocation = [System.IO.Path]::ChangeExtension($logFilePath, $newExtension)
-$sarifLog | ConvertTo-Json -depth 999 | Out-File $writeLocation
+$sarifLog | ConvertTo-Json -Compress -depth 999 | Out-File $writeLocation
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: This removes all white space and newlines. In general: favor readability over file size unless and until you have data demonstrating that file size is a problem. So revert this change.

**Comment by lgolding** [comment]
:clock1:

File: appveyor/Convert-FileUrisToGitHubUris.ps1 (hunk 2 above)
Comment (Rolstenhouse) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Fixed. I don't think I understood this distinction before.

File: appveyor/Convert-FileUrisToGitHubUris.ps1 (hunk 9 above)
Comment (Rolstenhouse) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: @rtaket had suggested to add the compression flag, will revert

File: appveyor/Convert-FileUrisToGitHubUris.ps1 (hunk 8 above)
Comment (Rolstenhouse) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Will tab

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 10]
```diff
@@ -16,17 +17,19 @@ else{
...
-      $replaceLength = $caseSensitiveUri.IndexOf($projectSlug)
-      if($replaceLength -ne -1){
-          $builder.Path = ($repoName, $repoCommit, $caseSensitiveUri.SubString($replaceLength+$projectSlug.Length+1) -join '/')
-          return $builder.ToString()
-      }
+function Rebase-Uri($originalUri){
+  if($originalURI){
+    $fileUriPrefix = "file:///"
+    if($originalUri.StartsWith($fileUriPrefix)){
+        $caseSensitiveUri = Get-CaseSensitivePath $originalUri.SubString($fileUriPrefix.Length
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Missing the closing paren. Does this even run? ;-)

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 11]
```diff
@@ -16,17 +17,19 @@ else{
...
+        $caseSensitiveUri = Get-CaseSensitivePath $originalUri.SubString($fileUriPrefix.Length 
+        $projectSlugIndex = $caseSensitiveUri.IndexOf($projectSlug)
+        if($projectSlugIndex -ne -1){
+            $builder.Path = ($repoName, $repoCommit, $caseSensitiveUri.SubString($projectSlugIndex+$projectSlug.Length+1) -join '/')
+            return $builder.ToString()
+        }
+    }
   }
-  Write-Error "Unable to rewrite URI"
+  Write-Host "Unable to rewrite URI: " $originalURI
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: $originalUri. Also on the line below. PS is case-insensitive for variable names, which is why you have to watch out for this yourself.

File: appveyor/Convert-FileUrisToGitHubUris.ps1 (hunk 7 above)
Comment (Rolstenhouse) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Fixed

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 12]
```diff
@@ -50,20 +53,24 @@ function Get-CaseSensitivePath($pathName){
...
-        $content.uri = Rebase-Uri $content.uri
-    }
-
-    if($content.parentKey){
-      $content.parentKey = Rebase-Uri $content.parentKey
+  if($sarifLog.runs[$i].files){
+    $sarifLog.runs[$i].files.PSObject.Properties | foreach-object{
+      $key = $_.Name
+      $value = $_.Value
+      if($key -and $content){
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: You changed the variable you assigned to, but not any of the _uses_ of that variable. This presumably doesn't run correctly.

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 13]
```diff
@@ -50,20 +53,24 @@ function Get-CaseSensitivePath($pathName){
...
+        if($content.uri){
+            $content.uri = Rebase-Uri $content.uri
+        }
+
+        if($content.parentKey){
+          $content.parentKey = Rebase-Uri $content.parentKey
+        }
+
+        $sarifLog.runs[$i].files.PSObject.Properties.Remove($key)
+        $rewrite = Rebase-Uri $key
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: `$rewrite` => `$rebasedUri`

File: appveyor/Convert-FileUrisToGitHubUris.ps1 (hunk 11 above)
Comment (Rolstenhouse) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Good catch

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 14]
```diff
@@ -16,17 +17,19 @@ else{
...
-      $replaceLength = $caseSensitiveUri.IndexOf($projectSlug)
-      if($replaceLength -ne -1){
-          $builder.Path = ($repoName, $repoCommit, $caseSensitiveUri.SubString($replaceLength+$projectSlug.Length+1) -join '/')
-          return $builder.ToString()
-      }
+function Rebase-Uri($originalUri){
+  if($originalURI){
+    $fileUriPrefix = "file:///"
+    if($originalUri.StartsWith($fileUriPrefix)){
+        $caseSensitiveUri = Get-CaseSensitivePath $originalUri.SubString($fileUriPrefix.Length
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Still missing closing paren.

Comment (Rolstenhouse) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Fixed

File: appveyor/Convert-FileUrisToGitHubUris.ps1 (hunk 13 above)
Comment (Rolstenhouse) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Multiple commits :/

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 15]
```diff
@@ -16,18 +17,20 @@ else{
...
-          return $builder.ToString()
-      }
+function Rebase-Uri($originalUri){
+  if($originalUri){
+    $fileUriPrefix = "file:///"
+    if($originalUri.StartsWith($fileUriPrefix)){
+        $caseSensitiveUri = Get-CaseSensitivePath $originalUri.SubString($fileUriPrefix.Length)
+        $projectSlugIndex = $caseSensitiveUri.IndexOf($projectSlug)
+        if($projectSlugIndex -ne -1){
+            $builder.Path = ($repoName, $repoCommit, $caseSensitiveUri.SubString($projectSlugIndex+$projectSlug.Length+1) -join '/')
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: >( [](start = 28, length = 1)

The closing paren is in the wrong place. You wanted:

```
$builder.Path = ($repoName, $repoCommit, $caseSensitiveUri.SubString($projectSlugIndex+$projectSlug.Length+1)) -join '/'
```

File: appveyor/Convert-FileUrisToGitHubUris.ps1 [hunk 16]
```diff
@@ -16,18 +17,20 @@ else{
...
+            return $builder.ToString()
+        }
+    }
   }
-  Write-Error "Unable to rewrite URI"
-  return $originalURI
+  Write-Host "Unable to rewrite Uri: " $originalUri
```
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: >Uri [](start = 32, length = 3)

:-) That casing convention applies to code. In written English, you continue to capitalize acronyms, _e.g._, URI.

File: appveyor/Convert-FileUrisToGitHubUris.ps1 (hunk 8 above)
Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Sorry, no, I meant "maybe there was a tab there, and that's why the indentation was wrong." Please use spaces :-)

---
In reply to: [71192075](https://github.com/Microsoft/sarif-sdk/pull/445#discussion_r71192075) [](ancestors = 71192075)

File: appveyor/Convert-FileUrisToGitHubUris.ps1 (hunk 15 above)
Comment (Rolstenhouse) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: It's written a correct way as is and functions properly. What's interesting is it works both ways

Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Ok, that tells me something I didn't know: the priority of the comma operator is higher than that of `-join`.

In that case, the outer set of parentheses should not be necessary:

```
$builder.Path = $repoName, $repoCommit, $caseSensitiveUri.SubString($projectSlugIndex+$projectSlug.Length+1) -join '/'
```

Comment (lgolding) [inline appveyor/Convert-FileUrisToGitHubUris.ps1]: Yes, tested this:
```
PS G:\Code\WinCComp\src> $a = "a"
PS G:\Code\WinCComp\src> $b = "b"
PS G:\Code\WinCComp\src> $c = "c"
PS G:\Code\WinCComp\src> $d = $a, $b, $c -join '/'
PS G:\Code\WinCComp\src> $d
a/b/c
```
